rav1e-by-gop/src/api.rs

```diff
@@ -0,0 +1,154 @@
+        let pool = self.pool.as_ref().map(|p| p.clone()).unwrap_or_else(|| {
+            let p = ThreadPoolBuilder::new()
+                .num_threads(self.threads)
+                .build()
+                .unwrap();
+            Arc::new(p)
+        });
+
+        Context {
+            pool,
+            base: self.base,
+            _pd: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<T: Pixel + DeserializeOwned> Context<T> {
+    pub fn single_pass_config(&self) -> rav1e::Config {
+        rav1e::config::Config::new()
+            .with_thread_pool(self.pool.clone())
+            .with_encoder_config(self.base)
+    }
+
+    pub fn encode_segment(
+        &self,
+        data: SegmentData,
+        progress_sender: ProgressSender,
+        segment_output_file: PathBuf,
+    ) -> Result<()> {
+        let mut progress = ProgressInfo::new(
+            rav1e::prelude::Rational {
+                num: self.base.time_base.den,
+                den: self.base.time_base.num,
+            },
+            data.compressed_frames.len(),
+            {
+                let mut kf = BTreeSet::new();
+                kf.insert(data.start_frameno);
+                kf
+            },
+            data.segment_no + 1,
+            data.next_analysis_frame,
+            None,
+        );
+        let _ = progress_sender.send(ProgressStatus::Encoding(Box::new(progress.clone())));
+
+        let mut ctx: rav1e::Context<T> = self
+            .single_pass_config()
+            .new_context()
+            .expect("Cannot setup the encoder");
+
+        self.pool.spawn(move || {
+            let mut source = Source {
+                compressed_frames: data.compressed_frames.into_iter().map(Rc::new).collect(),
+                sent_count: 0,
+            };
+            let _ = progress_sender.send(ProgressStatus::Encoding(Box::new(progress.clone())));
+
+            let mut output =
+                create_file_muxer(&segment_output_file).expect("Failed to create segment output");
+
+            loop {
+                match process_frame(&mut ctx, &mut source).expect("Cannot encode the frame") {
+                    ProcessFrameResult::Packet(packet) => {
+                        output.write_frame(
+                            packet.input_frameno as u64,
+                            packet.data.as_ref(),
+                            packet.frame_type,
+                        );
+                        progress.add_packet(*packet);
+                        let _ = progress_sender
+                            .send(ProgressStatus::Encoding(Box::new(progress.clone())));
+                    }
+                    ProcessFrameResult::NoPacket(_) => {
+                        // Next iteration
+                    }
+                    ProcessFrameResult::EndOfSegment => {
+                        output.flush().unwrap();
+                        let _ = progress_sender.send(ProgressStatus::Encoded(progress.segment_idx));
```

Comment:
The other day I introduced a [change](https://github.com/rust-av/rav1e-by-gop/commit/5b44354c225a462170f9705781c525ca6abe9972) that uses an in-memory muxer and writes the whole encoded buffer to a file at the end (here). I didn't actually measure how much that helps performance, or if at all, but is that something you think is reasonable to keep?

Reply:
On normal operating systems it wouldn't make a big difference since the files should just stay in cache, on the other hand I need to figure out if we could always enforce the wavefront scenario, so it would make sense to avoid a round trip over the kernel since we always know the first segment will be emitted among the firsts.

More about it on irc.